benchmarks/float8/utils.py

```diff
@@ -350,3 +336,326 @@ def get_gpu_kernel_gemm_time_s(f, *args, **kwargs):
+# around linear layers.
+#
+# The goal here is to set up microbenchmarks on the fusion of float8 
+# scaling/casting that are relevant to transformer models, and break
+# them up in such a way that we can easily reason about fusions regarding
+# each linear layer without having to dig through large traces / long log files.
+# There is some duplication here (attn_norm and ffn_norm appear in multiple
+# places), and that's ok.
+#
+# Simplifying assumptions (may be relaxed in the future):
+# 1. don't model fusion with rotary embeddings and repeat-transpose
+# 2. don't model fusion with SDPA
+#
+# Patterns:
+# 1. add -> attn_norm -> attn.{wq|wk|wv}
+#   - note: rotary embedding, repeat-transpose, SDPA not modeled for simplicity
+#   - note: model just one of (q, k, v) and assume the others are similar
+# 2. transpose-contiguous -> attn.wo -> add -> ffn_norm
+# 3. add -> ffn_norm -> ffn.w1 -> silu -> mul
+#   - note: assume that `add -> ffn_norm -> ffn.w3 -> mul` is covered by above
+# 4. silu -> mul -> ffn.w2 -> add -> attn_norm
+#   - note: in today's torchtitan codebase, the transformer blocks are compiled 
+#     separately, so `add` and `attn_norm` are not in the same graph
+#
+
+class AttnWQKVSubgraph(nn.Module):
+    def __init__(self, dim1, dim2):
+        super().__init__()
+        self.attn_norm = RMSNorm(dim1)
+        self.fc = nn.Linear(dim1, dim2, bias=False)
+
+    def forward(self, x, y):
+        x = x + y
+        x = self.attn_norm(x)
+        x = self.fc(x)
+        return x
+
+
+class AttnWOSubgraph(nn.Module):
+    def __init__(self, dim1, dim2):
+        super().__init__()
+        self.fc = nn.Linear(dim1, dim2, bias=False)
+        self.ffn_norm = RMSNorm(dim2)
+
+    def forward(self, x, y):
+        x = x.transpose(0, 1).contiguous()
+        x = self.fc(x)
+        x = x + y
+        x = self.ffn_norm(x)
+        return x
+
+
+class FFNW13Subgraph(nn.Module):
+    def __init__(self, dim1, dim2):
+        super().__init__()
+        self.ffn_norm = RMSNorm(dim1)
+        self.fc = nn.Linear(dim1, dim2, bias=False)
+
+    def forward(self, x, y):
+        x = x + y
+        x = self.ffn_norm(x)
+        x = self.fc(x)
+        x = F.silu(x)
+        x = x * x
+        return x
+
+
+class FFNW2Subgraph(nn.Module):
+    def __init__(self, dim1, dim2):
+        super().__init__()
+        self.fc = nn.Linear(dim1, dim2, bias=False)
+        self.attn_norm = RMSNorm(dim2)
+
+    def forward(self, x, y):
+        x = F.silu(x)
+        x = x * x
+        x = self.fc(x)
+        x = x + y
+        x = self.attn_norm(x)
+        return x
+
+
+class ModelType(enum.Enum):
+    LINEAR = "linear"
+    LN_LINEAR = "ln_linear"
+    NORM_FFN_NORM = "norm_ffn_norm"
+    NORM_FFN_NORM_SMALL = "norm_ffn_norm_small"
+    ATTN_WQKV_SUBGRAPH = "attn_wqkv_subgraph"
+    ATTN_WO_SUBGRAPH = "attn_wo_subgraph"
+    FFN_W13_SUBGRAPH = "ffn_w13_subgraph"
+    FFN_W2_SUBGRAPH = "ffn_w12_subgraph"
+
+
+def subgraph_name_to_subgraph_and_inputs(
```

Comment:
I think it would be nice if this was based off this had a few model sizes stamped out similar to 
https://github.com/pytorch-labs/ao_benchmarks/blob/8c5c95541198c1da2042d7028cab707e6541bb8a/experimental/smoothquant/debug/debug_llmi_blocks.py#L58-L80